PM Device Runtime: Introduce request/release reference APIs

bjarki-andreasen opened the pull request:
Applications migrating from using `pm_device_action_run()` to the reference counted `pm_device_runtime_get()`/`pm_device_runtime_put()` APIs are typically built around a pattern like:
```
/* resume if not already resumed */   
ret = pm_device_action_run(dev, PM_DEVICE_ACTION_RESUME);
if (ret == 0 || ret == -EALREADY) {
        return 0;
};

/* suspend if not already suspended */   
ret = pm_device_action_run(dev, PM_DEVICE_ACTION_SUSPEND);
if (ret == 0 || ret == -EALREADY) {
        return 0;
};
```
which allows multiple re-entrant paths to "force" the PM state if a device. This pattern is problematic if the reference counted APIs are used as it relies on balanced calls to get()/put(). To help applications migrate, and to provide an implementation of this pattern for applications to continue to use this pattern if they so wish, introduce the new APIs:
* `void pm_device_runtime_reference_init(struct pm_device_runtime_reference *ref);`
* `int pm_device_runtime_request(const struct device *dev,
			      struct pm_device_runtime_reference *ref);`
* `int pm_device_runtime_release(const struct device *dev,
			      struct pm_device_runtime_reference *ref);`
* `int pm_device_runtime_release_async(const struct device *dev,
				    struct pm_device_runtime_reference *ref,
				    k_timeout_t delay);`

These APIs let the application create references which can be "activated" and "deactivated", which use the reference counted APIs internally. The new pattern looks like:
```
/* reference used to balance reference counted calls internally  */
struct pm_device_runtime_reference ref;

pm_device_runtime_reference_init(&ref);

/* resume if not already resumed */   
return pm_device_runtime_request(dev, &ref);

/* suspend if not already suspended */   
return pm_device_runtime_release(dev, &ref);
```
The pattern is slightly different given `pm_device_runtime_release` does not ensure the device is actually suspended, this is by design as its based on `PM_DEVICE_RUNTIME` within which this is not possible to force. If applications where using `pm_device_action_run()` safely before, meaning no racing/conflicting calls to `pm_device_action_run()`, the new pattern should not introduce regressions :)

Like the `pm_device_runtime_get()/_put()` APIs, the structs, macros, and functions have stubs which completely optimize away if `CONFIG_PM_DEVICE_RUNTIME` is not enabled, so applications can be built with support for these new APIs with no ifdefs.

Includes bug fix from #94052

bjarki-andreasen-nordic committed:
pm: device_runtime: introduce pm_device_runtime_request/release()

Introduce the following APIs:

  - pm_device_runtime_reference_init()
  - pm_device_runtime_request()
  - pm_device_runtime_release()
  - pm_device_runtime_release_async()

and following macros

  - PM_DEVICE_RUNTIME_REFERENCE_INIT()
  - PM_DEVICE_RUNTIME_REFERENCE_DEFINE()

These APIs provide a request/release style interface built on top of
the reference counted pm_device_runtime APIs. The new APIs are useful
for applications migrating from using pm_device_action_run() to
PM_DEVICE_RUNTIME, especially replacing this pattern:

  int ret;

  ret = pm_device_action_run(dev, PM_DEVICE_ACTION_RESUME);
  if (ret == 0 || ret == -EALREADY) {
          return 0;
  };

Applications may also simply prefer the new APIs over the reference
counted ones, they can be a bit safer than reference counted APIs at
the cost of additional ROM/RAM.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
tests: pm: device_runtime_api: add pm_device_runtime_request/release

Extend device_runtime_api test suite to cover

  - pm_device_runtime_reference_init()
  - pm_device_runtime_request()
  - pm_device_runtime_release()
  - pm_device_runtime_release_async()
  - PM_DEVICE_RUNTIME_REFERENCE_DEFINE()

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

ceolin reviewed (commented):
@bjarki-andreasen I went through these additions and I can't still understand it. This is pretty much a lock around device runtime APIs, why we are adding this new API instead of addressing any issue in the device runtime and drivers ? The whole idea with PM enhancements is to make it simpler (and fix errors / inconsistencies) and I am not understanding how it will help.

I can see the current problem in drivers but we should get one API  / way to address it, if that is changing the current API, lets do it.

kartben commented on include/zephyr/pm/device_runtime.h:
```diff
@@ -221,6 +302,43 @@ static inline int pm_device_runtime_usage(const struct device *dev)
 	return -ENOSYS;
 }
 
+struct pm_device_runtime_reference {
+};
+
+#define PM_DEVICE_RUNTIME_REFERENCE_DEFINE(name) \
+	struct pm_device_runtime_reference name;
+
+static inline void pm_device_runtime_reference_init(struct pm_device_runtime_reference *ref)
+{
+	ARG_UNUSED(ref);
+}
+
+static inline int pm_device_runtime_request(const struct device *dev,
+					    struct pm_device_runtime_reference *ref)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(ref);
+	return 0;
+}
+
+static inline int pm_device_runtime_release(const struct device *dev,
+					    struct pm_device_runtime_reference *ref)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(ref);
+	return 0;
+}
+
+static inline int pm_device_runtime_release_async(const struct device *dev,
+						  struct pm_device_runtime_reference *ref,
+						  k_timeout_t delay)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(ref);
+	ARG_UNUSED(delay);
+	return 0;
+}
```
Of course! Sorry that I missed this.

decsny commented:
> > @bjarki-andreasen I went through these additions and I can't still understand it. This is pretty much a lock around device runtime APIs, why we are adding this new API instead of addressing any issue in the device runtime and drivers ? The whole idea with PM enhancements is to make it simpler (and fix errors / inconsistencies) and I am not understanding how it will help.
> > I can see the current problem in drivers but we should get one API / way to address it, if that is changing the current API, lets do it.
> 
> This API is meant to be used by applications to activate/deactivate **a** reference to a device. This is contrary to increasing a reference count, which then has to be tracked and balanced by the application. Its just a different pattern for adding/removing a reference. It does not lock a device any more or less than `pm_device_runtime_get/put` does, since it internally calls this API.
> 
> Applications use this pattern, so we should provide a well tested generic implementation of it, rather than having every application reimplement the following (simplified, excludes thread safety):
> 
> ```
> bool has_reference = false;
> 
> int enable_dev(void)
> {
>         if (has_reference) {
>                 return 0;
>         }
> 
>         if (pm_device_runtime_get(dev))
>                 return -EIO;
>         }
> 
>         has_reference = true;
>         return 0;
> }
> 
> int disable_dev(void)
> {
>         if (!has_reference) {
>                 return 0;
>         }
> 
>         if (pm_device_runtime_put(dev))
>                 return -EIO;
>         }
> 
>         has_reference = false;
>         return 0;
> }
> ```
> 
> this is common in applications, especially ones still using pm_device_action_run(). We should support it properly to make transitions easier.

Now I think I understand, this is actually also a problem for drivers, and pm consumers that are still using state locks directly instead of using pm runtime yet, for example https://github.com/zephyrproject-rtos/zephyr/blob/c6f31346d24f2839d8072d37acda5ed55ec21706/drivers/serial/uart_mcux_flexcomm.c#L107-L145

I imagine just like the point of that boolean in that driver, that this helps avoid race conditions and bugs caused by some code accidentally referencing a device twice when it didn't mean to

decsny commented on subsys/pm/device_runtime.c:
```diff
@@ -601,6 +601,109 @@ int pm_device_runtime_usage(const struct device *dev)
+	ref->active = false;
+}
+
+static bool reference_trylock(struct pm_device_runtime_reference *ref)
+{
+	if (k_is_in_isr() && k_sem_take(&ref->lock, K_NO_WAIT)) {
+		return false;
+	}
+
+	(void)k_sem_take(&ref->lock, K_FOREVER);
+	return true;
+}
+
+static void reference_unlock(struct pm_device_runtime_reference *ref)
+{
+	k_sem_give(&ref->lock);
+}
```
It seems like a semaphore might be overkill here, if it's just guarding access to the bool as a critical section, could it just be a spinlock instead? Or are you thinking that the `pm_device_runtime_get` will take a long time due to traversing domain callbacks?

bjarki-andreasen-nordic committed:
drivers: led: led_pwm: refactor to use pm_device_runtime

Refactor led_pwm device driver to use pm_device_runtime the
pm_device_driver_init helper. Additionally, include base.yaml
in the devicetree binding for pwm-leds as it contains the
zephyr,pm-device-runtime-auto property.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

decsny reviewed (commented):
I feel though it is not clear based on doxygen that these new API are not really meant to be used on a device at the same time as the other pm device APIs

tmleman commented on drivers/led/led_pwm.c:
```diff
@@ -68,66 +70,117 @@ static int led_pwm_set_brightness(const struct device *dev,
+			err = pm_device_runtime_release(led->dev, ref);
+		}
+
+		if (err) {
+			LOG_DBG("PWM %p %s failed (err = %d)", led->dev, request_str, ret);
+			ret = -EIO;
 		}
```
Using `ret` instead of `err`. Should be:
```c
if (err) {
	LOG_DBG("PWM %p %s failed (err = %d)", led->dev, request_str, err);
	ret = -EIO;
}
```

bjarki-andreasen commented:
sorry, messed up a bit on that force push, 2 sec
